lib/auth/register.go

```diff
@@ -740,23 +756,89 @@ func registerUsingAzureMethod(
 		}
 
 		return &proto.RegisterUsingAzureMethodRequest{
-			RegisterUsingTokenRequest: &types.RegisterUsingTokenRequest{
-				Token:                token,
-				HostID:               params.ID.HostUUID,
-				NodeName:             params.ID.NodeName,
-				Role:                 params.ID.Role,
-				AdditionalPrincipals: params.AdditionalPrincipals,
-				DNSNames:             params.DNSNames,
-				PublicTLSKey:         params.PublicTLSKey,
-				PublicSSHKey:         params.PublicSSHKey,
-			},
-			AttestedData: ad,
-			AccessToken:  accessToken,
+			RegisterUsingTokenRequest: registerUsingTokenRequestForParams(token, params),
+			AttestedData:              ad,
+			AccessToken:               accessToken,
 		}, nil
 	})
 	return certs, trace.Wrap(err)
 }
 
+// registerUsingTPMMethod is used to register using the TPM join method. It
+// is able to register through a proxy or through the auth server directly.
+func registerUsingTPMMethod(
+	ctx context.Context,
+	client joinServiceClient,
+	token string,
+	params RegisterParams,
+) (*proto.Certs, error) {
+	log := slog.Default()
+
+	initReq := &proto.RegisterUsingTPMMethodInitialRequest{
+		JoinRequest: registerUsingTokenRequestForParams(token, params),
+	}
+
+	attestation, close, err := tpm.Attest(ctx, log)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer func() {
+		if err := close(); err != nil {
+			log.WarnContext(ctx, "Failed to close TPM", "error", err)
+		}
+	}()
+
+	initReq.AttestationParams = tpm.AttestationParametersToProto(
+		attestation.AttestParams,
+	)
+	// Get the EKKey or EKCert. We want to prefer the EKCert if it is available
+	// as this is signed by the manufacturer.
+	switch {
+	case attestation.Data.EKCert != nil:
+		log.DebugContext(
+			ctx,
+			"Using EKCert for TPM registration",
+			slog.String("ekcert_serial", attestation.Data.EKCert.SerialNumber),
```

Comment:
nit: I don't think we are gaining much by using an explicit Attr here?

```suggestion
			"ekcert_serial", attestation.Data.EKCert.SerialNumber,
```

Reply:
Fair enough, your call.